internal/controller/cluster_monitoring_controller.go

```diff
@@ -126,43 +126,14 @@ func (r *ClusterMonitoringReconciler) reconcile(ctx context.Context, cluster *cl
 	// Add finalizer first if not set to avoid the race condition between init and delete.
 	// Note: Finalizers in general can only be added when the deletionTimestamp is not set.
 	if !controllerutil.ContainsFinalizer(cluster, monitoring.MonitoringFinalizer) {
-		// We use a patch rather than an update to avoid conflicts when multiple controllers are adding their finalizer to the ClusterCR
-		// We use the patch from sigs.k8s.io/cluster-api/util/patch to handle the patching without conflicts
-		logger.Info("adding finalizer", "finalizer", monitoring.MonitoringFinalizer)
-		patchHelper, err := patch.NewHelper(cluster, r.Client)
-		if err != nil {
-			return ctrl.Result{}, errors.WithStack(err)
-		}
-		controllerutil.AddFinalizer(cluster, monitoring.MonitoringFinalizer)
-		if err := patchHelper.Patch(ctx, cluster); err != nil {
-			logger.Error(err, "failed to add finalizer", "finalizer", monitoring.MonitoringFinalizer)
-			return ctrl.Result{}, errors.WithStack(err)
-		}
-		logger.Info("added finalizer", "finalizer", monitoring.MonitoringFinalizer)
-		return ctrl.Result{}, nil
+		return r.addFinalizer(ctx, cluster)
 	}
 
 	// Management cluster specific configuration
 	if cluster.Name == r.ManagementCluster.Name {
-		// If monitoring is enabled as the installation level, configure the monitoring stack, otherwise, tear it down.
-		if r.MonitoringConfig.Enabled {
-			err = r.HeartbeatRepository.CreateOrUpdate(ctx)
-			if err != nil {
-				logger.Error(err, "failed to create or update heartbeat")
-				return ctrl.Result{RequeueAfter: 5 * time.Minute}, nil
-			}
-
-			err = r.MimirService.ConfigureMimir(ctx)
-			if err != nil {
-				logger.Error(err, "failed to configure mimir")
-				return ctrl.Result{RequeueAfter: 5 * time.Minute}, nil
-			}
-		} else {
-			err = r.tearDown(ctx)
-			if err != nil {
-				logger.Error(err, "failed to tear down the monitoring stack")
-				return ctrl.Result{RequeueAfter: 5 * time.Minute}, nil
-			}
+		result := r.reconcileManagementCluster(ctx)
```

Comment:
I don't think it makes sense to return a pointer here.
I would suggest that you only return an error and if the err is not nil then you return `ctrl.Result{RequeueAfter: 5 * time.Minute}` as this is what we return on all 3 error cases anyway :D

Reply:
That does not address this though :(